// @iota/dapp-kit/dist/cjs/hooks/useGetDefaultIotaName.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/hooks/useGetDefaultIotaName.ts
var useGetDefaultIotaName_exports = {};
__export(useGetDefaultIotaName_exports, {
  useGetDefaultIotaName: () => useGetDefaultIotaName
});
module.exports = __toCommonJS(useGetDefaultIotaName_exports);
var import_client2 = require("@iota/iota-sdk/client");
var import_react_query = require("@tanstack/react-query");
var import_iota_names_sdk = require("@iota/iota-names-sdk");

// src/hooks/useIotaClient.ts
var import_react2 = require("react");

// src/components/IotaClientProvider.tsx
var import_client = require("@iota/iota-sdk/client");
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var IotaClientContext = (0, import_react.createContext)(null);
var DEFAULT_NETWORKS = {
  localnet: { url: (0, import_client.getNetwork)(import_client.Network.Localnet)?.url }
};

// src/hooks/useIotaClient.ts
function useIotaClientContext() {
  const iotaClient = (0, import_react2.useContext)(IotaClientContext);
  if (!iotaClient) {
    throw new Error(
      "Could not find IotaClientContext. Ensure that you have set up the IotaClientProvider"
    );
  }
  return iotaClient;
}

// src/hooks/useGetDefaultIotaName.ts
var import_react3 = require("react");
var import_graphql = require("@iota/iota-sdk/graphql");
function useGetDefaultIotaName(address, iotaNamesEnabled) {
  const iotaContext = useIotaClientContext();
  const network = (0, import_client2.getNetwork)(iotaContext.network);
  const iotaNamesClient = (0, import_react3.useMemo)(() => {
    const iotaGraphQLClient = new import_graphql.IotaGraphQLClient({
      url: network.graphql
    });
    return new import_iota_names_sdk.IotaNamesClient({
      graphQlClient: iotaGraphQLClient,
      network: network.id
    });
  }, [network.id]);
  const iotaNamesSupported = !!iotaNamesClient.config;
  return (0, import_react_query.useQuery)({
    queryKey: ["iota-name", "default-name", address],
    queryFn: async () => {
      if (!address) return null;
      const name = await iotaNamesClient?.getDefaultName(address);
      return name ? (0, import_iota_names_sdk.normalizeIotaName)(name) : name;
    },
    enabled: !!iotaNamesClient && !!address && iotaNamesEnabled && iotaNamesSupported,
    staleTime: 1e3 * 60 * 10
  });
}
