// @iota/dapp-kit/dist/esm/hooks/useGetDefaultIotaName.js
// src/hooks/useGetDefaultIotaName.ts
import { getNetwork as getNetwork2 } from "@iota/iota-sdk/client";
import { useQuery } from "@tanstack/react-query";
import { IotaNamesClient, normalizeIotaName } from "@iota/iota-names-sdk";

// src/hooks/useIotaClient.ts
import { useContext } from "react";

// src/components/IotaClientProvider.tsx
import { getNetwork, isIotaClient, Network, IotaClient } from "@iota/iota-sdk/client";
import { createContext, useMemo, useState } from "react";
import { jsx } from "react/jsx-runtime";
var IotaClientContext = createContext(null);
var DEFAULT_NETWORKS = {
  localnet: { url: getNetwork(Network.Localnet)?.url }
};

// src/hooks/useIotaClient.ts
function useIotaClientContext() {
  const iotaClient = useContext(IotaClientContext);
  if (!iotaClient) {
    throw new Error(
      "Could not find IotaClientContext. Ensure that you have set up the IotaClientProvider"
    );
  }
  return iotaClient;
}

// src/hooks/useGetDefaultIotaName.ts
import { useMemo as useMemo2 } from "react";
import { IotaGraphQLClient } from "@iota/iota-sdk/graphql";
function useGetDefaultIotaName(address, iotaNamesEnabled) {
  const iotaContext = useIotaClientContext();
  const network = getNetwork2(iotaContext.network);
  const iotaNamesClient = useMemo2(() => {
    const iotaGraphQLClient = new IotaGraphQLClient({
      url: network.graphql
    });
    return new IotaNamesClient({
      graphQlClient: iotaGraphQLClient,
      network: network.id
    });
  }, [network.id]);
  const iotaNamesSupported = !!iotaNamesClient.config;
  return useQuery({
    queryKey: ["iota-name", "default-name", address],
    queryFn: async () => {
      if (!address) return null;
      const name = await iotaNamesClient?.getDefaultName(address);
      return name ? normalizeIotaName(name) : name;
    },
    enabled: !!iotaNamesClient && !!address && iotaNamesEnabled && iotaNamesSupported,
    staleTime: 1e3 * 60 * 10
  });
}
export {
  useGetDefaultIotaName
};
