// @iota/iota-names-sdk/dist/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  ALLOWED_METADATA: () => import_constants.ALLOWED_METADATA,
  GRACE_PERIOD_MS: () => import_constants.GRACE_PERIOD_MS,
  IotaNamesClient: () => import_iota_names_client.IotaNamesClient,
  IotaNamesTransaction: () => import_iota_names_transaction.IotaNamesTransaction,
  MIN_LABEL_SIZE: () => import_constants.MIN_LABEL_SIZE,
  getConfigType: () => import_helpers.getConfigType,
  getNameRegistrationType: () => import_helpers.getNameRegistrationType,
  getNameType: () => import_helpers.getNameType,
  getPricelistConfigType: () => import_helpers.getPricelistConfigType,
  getRenewalPricelistConfigType: () => import_helpers.getRenewalPricelistConfigType,
  getSubnameRegistrationType: () => import_helpers.getSubnameRegistrationType,
  isNestedSubname: () => import_helpers.isNestedSubname,
  isSubname: () => import_helpers.isSubname,
  isValidIotaName: () => import_utils.isValidIotaName,
  normalizeIotaName: () => import_utils.normalizeIotaName,
  packages: () => import_constants.packages,
  validateIotaName: () => import_utils.validateIotaName,
  validateIotaSubname: () => import_utils.validateIotaSubname,
  validateYears: () => import_helpers.validateYears
});
module.exports = __toCommonJS(index_exports);
var import_iota_names_client = require("./iota-names-client.js");
var import_iota_names_transaction = require("./iota-names-transaction.js");
var import_constants = require("./constants.js");
var import_helpers = require("./helpers.js");
var import_utils = require("./utils.js");
